src/unix/thread.c

```diff
@@ -500,7 +504,102 @@ int uv_sem_trywait(uv_sem_t* sem) {
   return 0;
 }
 
-#else /* !(defined(__APPLE__) && defined(__MACH__)) */
+#elif defined(__GLIBC__) && __GLIBC_MINOR__ < 21
+
+/* Hack around https://sourceware.org/bugzilla/show_bug.cgi?id=12674
+ * by providing a custom implementation for glibc < 2.21 in terms of other
+ * concurrency primitives.
+ * Refs: https://github.com/nodejs/node/issues/19903 */
+
+/* To preserve ABI compatibility, we treat the uv_sem_t as storage for
+ * a pointer to the actual struct we're using underneath. */
+
+
+typedef struct uv_semaphore_s {
+  uv_mutex_t mutex;
+  uv_cond_t cond;
+  unsigned int value;
+} uv_semaphore_t;
+
+
+int uv_sem_init(uv_sem_t* sem_, unsigned int value) {
+  int err;
+  uv_semaphore_t* sem;
+
+  sem = malloc(sizeof(uv_semaphore_t));
+  if (sem == NULL)
+    return UV_ENOMEM;
+
+  if ((err = uv_mutex_init(&sem->mutex)) != 0) {
+    free(sem);
+    return err;
+  }
+
+  if ((err = uv_cond_init(&sem->cond)) != 0) {
+    uv_mutex_destroy(&sem->mutex);
+    free(sem);
+    return err;
+  }
+
+  assert(sizeof(uv_sem_t) >= sizeof(uv_semaphore_t*));
+  sem->value = value;
+  *(uv_semaphore_t**) sem_ = sem;
```

Comment:
You're welcome to turn `uv_sem_t` into an union if you think that's neater.  It wouldn't be an API change (opaque type) and not an ABI break either.

Reply:
I can do that if you like, but tbh I think this is fine as-is for now.

Reply:
Okay, that's fine.